src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1633,46 +1606,25 @@ _mongoc_topology_push_server_session (mongoc_topology_t *topology,
    ENTRY;
 
    bson_mutex_lock (&topology->mutex);
-
    timeout = topology->description.session_timeout_minutes;
    loadbalanced = topology->description.type == MONGOC_TOPOLOGY_LOAD_BALANCED;
+   bson_mutex_unlock (&topology->mutex);
 
-   /* start at back of queue and reap timed-out sessions */
```

Comment:
This does diverge from the driver specification's [Algorithm to return a ServerSession instance to the server session pool](https://github.com/mongodb/specifications/blob/df6be82f865e9b72444488fd62ae1eb5fca18569/source/sessions/driver-sessions.rst#algorithm-to-return-a-serversession-instance-to-the-server-session-pool)

> Before returning a server session to the pool a driver MUST first check the server session pool for server sessions at the back of the queue that are about to expire (meaning they will expire in less than one minute). A driver MUST stop checking server sessions once it encounters a server session that is not about to expire. Any server sessions found that are about to expire are removed from the end of the queue and discarded (or allowed to be garbage collected)

Would adding this functionality back nullify the performance improvement gained here?

Reply:
It would not entirely negate the performance improvement, since the pop operations are very fast and multiple threads can simultaneously pop.

However, I question the need for that to be in the spec, as it is more of an implementation concern than a behavioral change.

I presume it was specified that way to prevent the pool from growing with expired sessions ad infinitum, but I don't think the design of the pool as a stack would ever allow such unbounded growth.